drivers: entropy: stm32: Add power management constraints

wcappelle opened the pull request:
Add power managment constraints to the entropy driver.
This prevents the hardware block to lose it's clock when going into any stop mode of the cpu, which would cause the clock error flag to be set while filling the pool.

This is a part of #42374

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -476,39 +457,33 @@ static int entropy_stm32_rng_init(const struct device *dev)
-
-#if DT_INST_NODE_HAS_PROP(0, health-test-config)
-#if DT_INST_NODE_HAS_PROP(0, health-test-magic)
+#if DT_INST_NODE_HAS_PROP(0, health - test - config)
+#if DT_INST_NODE_HAS_PROP(0, health - test - magic)
```
Seems clang formatting is broken.

wcappelle replied:
oh, good catch :-o

wcappelle replied:
euh, I think this is an error as well right? all properties are converted from - to _ right?

wcappelle replied:
@erwango 
just tested it, and never worked afaik. I can fix them properly with the '_', but probly this needs a new PR?

erwango replied:
ok, txs. Yes a new PR (+issue) would be great so we can put this in DV3.0.

wcappelle replied:
done in #42386

wcappelle replied:
reverted this change, and let #42386 fix this

gmarull replied:
Actually, I think these DT macros are broken, they should be using `_`:

```
#if DT_INST_NODE_HAS_PROP(0, health_test_config)
```

wcappelle replied:
yes, see previous comments. I've fixed them in #42386

gmarull commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -111,6 +112,24 @@ static int entropy_stm32_got_error(RNG_TypeDef *rng)
 }
 
+#ifdef CONFIG_PM
+static void entropy_stm32_pm_constraint_set(struct entropy_stm32_rng_dev_data *data)
+{
+	if (!data->pm_constraint_on) {
+		data->pm_constraint_on = true;
+		pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
+	}
+}
+
+static void entropy_stm32_pm_constraint_release(struct entropy_stm32_rng_dev_data *data)
+{
+	if (data->pm_constraint_on) {
+		data->pm_constraint_on = false;
+		pm_constraint_release(PM_STATE_SUSPEND_TO_IDLE);
+	}
+}
```
What is the point of these wrapper functions? If the driver is well-behaved, it will never call a release before a previous set. API is reference counted, so it is fine to call set multiple times as soon as release is later called the same times.

wcappelle replied:
I just looked to other implementations like stm32 uart driver. It's more a safeguard indeed to not call the set more than intended

gmarull replied:
This needs a justification, constraint API can be safely called more than once as soon as set/release order is respected.

wcappelle replied:
yes, but the pm constraint api is not tracking the count per driver which is calling, so let's say there is a mistake in this driver, it would release the constraint of another driver. I think the safegaurd doesn't hurt, but if you don't agree, it's easy to remove.

wcappelle replied:
that's clear, I will rework this

wcappelle replied:
done

wcappelle commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -88,8 +86,7 @@ struct entropy_stm32_rng_dev_data {
 static const struct entropy_stm32_rng_dev_cfg entropy_stm32_rng_config = {
-	.pclken	= { .bus = DT_INST_CLOCKS_CELL(0, bus),
-		    .enr = DT_INST_CLOCKS_CELL(0, bits) },
+	.pclken = { .bus = DT_INST_CLOCKS_CELL(0, bus), .enr = DT_INST_CLOCKS_CELL(0, bits) },
```
yeah, it doesn't add up to the readability indeed..
I will revert the clang format commit since it's more question to debate than it helps it seems :-).

I'm really dissapointed (as you are probably) in the benefit of the current clang configuration as it is never helping to pass a review for code style. I think it's a real pain for new developers and reviewers especially since both lose time on something which can be done more automatically. Is there any issue for this, or should I/you create one?

wcappelle replied:
reverted

gmarull replied:
I agree, Zephyr should take this more seriously. In year 2022 we should no longer be discussing code formatting but rather say “please run this auto-formatter”. It's an area that IMO is crucial for a project to scale well. I've been working recently on simplifying `clang-format`, see https://github.com/zephyrproject-rtos/zephyr/pull/41307.

Commit:
drivers: entropy: stm32: Add power management constraints

Add power management constraints to the entropy driver.
This prevents the hardware block to lose it's clock when
going into any stop mode of the cpu, which would cause the
clock error flag to be set while filling the pool.

Signed-off-by: Wouter Cappelle <wouter.cappelle@crodeon.com>

wcappelle commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -65,12 +65,10 @@ struct rng_pool {
 
-BUILD_ASSERT((CONFIG_ENTROPY_STM32_ISR_POOL_SIZE &
-	      (CONFIG_ENTROPY_STM32_ISR_POOL_SIZE - 1)) == 0,
+BUILD_ASSERT((CONFIG_ENTROPY_STM32_ISR_POOL_SIZE & (CONFIG_ENTROPY_STM32_ISR_POOL_SIZE - 1)) == 0,
```
reverted

wcappelle commented:
@erwango , @gmarull Thanks for the review, I've updated the requested changes & reverted the clang formatting as it didn't do it quite right. 
Could you find some time to rereview?

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -492,6 +495,8 @@ static int entropy_stm32_rng_init(const struct device *dev)
 	LL_RNG_SetHealthConfig(dev_data->rng, DT_INST_PROP(0, health-test-config));
 #endif
 
+	pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
```
This one totally prevents the use of PM, right ?

wcappelle replied:
yes, for a small amount of time. At startup the rng pool is filled, so the clock needs to be on for that time. It's stopped when the rng pool is filled (in the isr)

wcappelle replied:
Should I add a comment to explain the flow?

erwango replied:
It would be helpful, txs

gmarull reviewed (dismissed):
LGTM, just one comment: I see 2 `pm_constraint_set` calls and 1 `pm_constraint_release` call, I suppose the driver takes care of sequencing, e.g. never call 2 sets and 1 release, etc. (or API is used in a way that this is prevented)

wcappelle commented:
> LGTM, just one comment: I see 2 `pm_constraint_set` calls and 1 `pm_constraint_release` call, I suppose the driver takes care of sequencing, e.g. never call 2 sets and 1 release, etc. (or API is used in a way that this is prevented)
Well, I think you might be right. with reverting the flag, I think the pm_constraint_set can indeed be set multiple times now before releasing. Let me check furter/fix that

wcappelle commented:
@gmarull , i've fixed the possible double set.

wcappelle commented:
@erwango , I've added the comment. Do you have and further review comments?
